src/gluonts/model/n_beats/_network.py

```diff
@@ -508,19 +517,37 @@ def __init__(
                     )
 
     # noinspection PyMethodOverriding,PyPep8Naming
-    def hybrid_forward(self, F, past_target: Tensor, future_target: Tensor):
+    def hybrid_forward(
+        self,
+        F,
+        past_target: Tensor,
+        past_observed_values: Tensor,
+        future_target: Optional[Tensor],
+        future_observed_values: Optional[Tensor],
+    ):
+
+        past_target, scale = self.scaler(past_target, past_observed_values)
+
         if len(self.net_blocks) == 1:  # if first block is also last block
-            return self.net_blocks[0](past_target)
+            forecast = self.net_blocks[0](past_target)
         else:
             backcast, forecast = self.net_blocks[0](past_target)
-            backcast = past_target - backcast
+            backcast = (past_target - backcast) * past_observed_values
             # connect regular blocks (all except last)
             for i in range(1, len(self.net_blocks) - 1):
                 b, f = self.net_blocks[i](backcast)
-                backcast = backcast - b
+                backcast = (backcast - b) * past_observed_values
                 forecast = forecast + f
             # connect last block
-            return forecast + self.net_blocks[-1](backcast)
+            forecast = forecast + self.net_blocks[-1](backcast)
+
+        forecast = F.broadcast_mul(forecast, scale)
+
+        return (
+            forecast
+            if future_observed_values is None
+            else forecast * future_observed_values
```

Comment:
Maybe an idea: just return the original (unmasked) `forecast` tensor here, and let the caller mask it where needed (e.g. the training network). Could that work?

In fact: only masking the forecast is probably not enough, and the corresponding target values which are compared in the loss should be masked too. Or otherwise, one can mask the loss values directly. (But I haven't checked in the paper if this is what the authors mean -- I hope so)

With these changes, the `future_` inputs could be removed from this block probably.

Reply:
I think it's a good idea, I moved the masking inside the loss functions. I followed the [authors' implementation](https://github.com/ElementAI/N-BEATS/blob/04f56c4ca4c144071b94089f7195b1dd606072b0/common/torch/losses.py#L45).
Still did not manage to remove future imports (did not even try since the tests were failing before).